Add sparse marlin AQT layout

Diogo-V opened the pull request:
## Summary
Introduces a Layout for AQT class called [Sparse Marlin 2:4](https://github.com/IST-DASLab/Sparse-Marlin). 

## E2E Results
<img width="952" alt="Screenshot 2024-09-06 at 8 57 06 AM" src="https://github.com/user-attachments/assets/50701029-d7e4-4fb5-ae13-e08ed5528c4c">

## Tests
test/sparsity/test_marlin.py

## Considerations for reviewers
1. To solve the issue with `torch.compile` I had to covert the section of [here](https://github.com/Diogo-V/ao/blob/f52925cb47c95eec95b1725a933fca99a8473d9d/torchao/sparsity/marlin/__init__.py#L239) and [here](https://github.com/Diogo-V/ao/blob/f52925cb47c95eec95b1725a933fca99a8473d9d/torchao/sparsity/marlin/__init__.py#L272) from numpy to torch. If I used torch.uint32, it would throw an error of `rshift_cpu not supported for uint32`. Using `torch.int64` seems to not cause any problems
2. I was able to completely remove the dependency from numpy, so we won't need to add it as a dependency

## Notes
* Closes #549 
* Since there is a deadline for the torchAO announcement tomorrow and, in the case that this work was to be part of it, I will be around today to work on any changes that this PR might need before it gets merged in.

Feel free to let me know if there is anything that needs to be refactored!

msaroufim commented on torchao/csrc/cuda/sparse_marlin/base.h:
```diff
@@ -0,0 +1,49 @@
+/*
```
Make sure to also create a test here https://github.com/pytorch/ao/tree/main/test for the cuda kernel and a test for an end to end flow https://github.com/pytorch/ao/tree/main/test/sparsity

msaroufim commented:
so you should have perms to trigger CI, keep in mind that CI wont run if you have merge conflicts like now

jcaip commented on torchao/sparsity/marlin.py:
```diff
@@ -0,0 +1,443 @@
+# This function converts dense matrix into sparse semi-structured
+# representation, producing "compressed" matrix, in the layout used by
+# CUTLASS backend, and corresponding metadata matrix.
+def _sparse_semi_structured_from_dense_cutlass(dense):
```
I believe that this are the same functions that are available here: https://github.com/pytorch/pytorch/blob/main/torch/sparse/_semi_structured_conversions.py

So we don't have to copy them over.

Diogo-V replied:
It seems some of them have small differences:
- https://www.diffchecker.com/ZWzCCGen/
- https://www.diffchecker.com/rIyyxGDx/
- https://www.diffchecker.com/kUzQSdpR/

Do you want me to leave it as is or should I change anything?

jcaip replied:
cc @alexsamardzic I'm assuming they may have changed some stuff for int4 support. n00b question here, would it make sense to upstream some of the changes that they made?

@Diogo-V let's leave it for now, we can revisit once Alex responds.

alexsamardzic replied:
Yes - it would be good to have this upstream, but it would be good then also to have corresponding tests in `test/test_sparse_semi_structured.py` extended accordingly.

alexsamardzic replied:
I'll make a note; if taking on this, I'd also fully extend CUTLASS variant of SparseSemiStructuredTensor for int4 support.  Am at the moment working on stuff related to mixed data types MM, so can't promise exactly if/when I'll have cycles for this - sorry about that.

jcaip reviewed (commented):
Looks like this is coming along nicely @Diogo-V ! Left a couple comments

BTW, once you get the subclass hooked up we should try this on SAM:
https://github.com/pytorch/ao/blob/main/torchao/_models/sam/eval_combo.py#L286

I'm putting together a collection of ViT sparsification results for our PTC poster and would love to add your result and give you a shout out. 

Also, feel free to @ me, I may not be the most responsive otherwise.

Diogo-V commented:
@jcaip - Thank you for the comments! I am going to continue working on this today and will take care of them. 
I think I have most things hooked together. There seems to be a weird bug somewhere but I will grind through it.

Once we merge this, I’d be happy to add it to SAM and run the benchmarks!

Also, feel free to let me know if there is anything that I can help out on for the PTC poster. I managed to finish mine earlier this week and so I have a bit more time 🙂

Diogo-V commented:
> so you should have perms to trigger CI, keep in mind that CI wont run if you have merge conflicts like now

Awesome! I am going to resolve them and try it out once I am at the computer

Diogo-V commented on test/sparsity/test_marlin.py:
```diff
@@ -0,0 +1,46 @@
+
+class TestQuantSparseMarlin(TestCase):
+
+    @pytest.mark.skipif(not TORCH_VERSION_AFTER_2_3, reason="pytorch 2.3+ feature")
```
Done

Diogo-V commented on torchao/dtypes/affine_quantized_tensor.py:
```diff
@@ -490,6 +497,95 @@ def from_plain(
+        return int_data_expanded, scales_expanded, self.zero_point
+
+    @classmethod
+    def from_plain(
```
Yeah, its for an override here: https://github.com/pytorch/ao/blob/b523f9f9e15b6fb80d10f585d9cf45e0c5e4d10e/torchao/dtypes/affine_quantized_tensor.py#L50

I can add a function called `from_dense` that would internally call `from_plain` to keep it consistent:
```py

@classmethod
def from_dense(cls, ...):
    return cls.from_plain(...)

```

If you have any other suggestions, let me know

jcaip replied:
No that's fine as is then. I think jerry probably updated the semi-structured sparse one as well and I didn't notice.

jerryzh168 replied:
yeah we are using `from_plain` to register the constructor for layout tensor in the table: https://github.com/pytorch/ao/blob/227d4bfece6bb00f0365642ad1d7afd509079ba4/torchao/dtypes/utils.py#L119

Diogo-V commented on torchao/sparsity/marlin.py:
```diff
@@ -0,0 +1,443 @@
+    return (cols_maj * m * interleave + dst_rows * interleave + cols_min).view(-1)
+
+
+def _mask_creator(tensor):
```
Done

jcaip commented on wip_test_llama2.py:
```diff
@@ -0,0 +1,24 @@
+from transformers import AutoTokenizer, LlamaForCausalLM
+
+device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+name = "meta-llama/Llama-2-7b-hf"
```
You can use `"nm-testing/SparseLlama-3-8B-pruned_50.2of4"` as a checkpoint

jcaip commented on torchao/dtypes/affine_quantized_tensor.py:
```diff
@@ -517,6 +529,153 @@ def from_plain(
+
+# Marlin Sparse op dispatch registration 
+@MarlinSparseAQTLayout.implements(aten.detach.default)
+def block_sparse_detach(func, types, args, kwargs):
```
nit: I think @jerryzh168 has been using `def _(` for all of these, you should do the same.

Diogo-V commented:
Hi, I did a bit more work on this PR this last few days and, if possible, I would like to ask for another preliminary review on it. It is not yet finished because I am still having a bug somewhere that is causing llama2 to output some weird text (I am using llama2 to validate everything E2E). Will be working on it the next couple of days:

<img width="1165" alt="image" src="https://github.com/user-attachments/assets/ed4a1c73-2646-45df-aef4-b91b0403c3e1">

I added a file in root of the project called `wip_test_llama.py` just to make it easier for you in case you want to check it. This file will be removed before the PR is merged.

The main difference from last time is that instead of using the code from the [original repo](https://github.com/IST-DASLab/Sparse-Marlin/tree/main) I went with the code that they have in the [nm-vllm repo](https://github.com/neuralmagic/nm-vllm/tree/main). Three reasons for it: the code was easier to follow, had better APIs and their kernel implementation also supports fp8 making it fairly straightforward to add support for it in the future if we want to.

There are two parts of the code of which I am not confident if it is correct. To make it easier for you to look for them, I added a comment above the code section with the text `NOTE(reviewers)`:
1. In [here](https://github.com/Diogo-V/ao/blob/7f9c65a9c49951d0b42d8d2d6294ee925fa5180b/torchao/dtypes/affine_quantized_tensor.py#L621), I am not sure if I should transpose it and/or if it is even correct how I did it. It was done to match the `in_features, out_features` format that was in the original repo and because I was having some trouble with the shapes 😅 
2. In [here](https://github.com/Diogo-V/ao/blob/7f9c65a9c49951d0b42d8d2d6294ee925fa5180b/torchao/dtypes/affine_quantized_tensor.py#L1058) I am not sure if I am handling the reshape of the input tensor correctly. 

Let me know if you find something that needs refactoring!

Aside from that, I have a question:
* There is a [section of code](https://github.com/Diogo-V/ao/blob/7f9c65a9c49951d0b42d8d2d6294ee925fa5180b/torchao/sparsity/marlin/__init__.py#L243) that is using numpy. I tried converting it to torch but it was throwing the error `Promotion for uint16, uint32, uint64 types is not supported, attempted to promote UInt32 and Int`. Is there a way for me to write this part of the code in native torch that would work? If not, then we will probably need to add numpy as a dependency to torchao before this gets merged. How should I proceed on this?

tagging for visibility: @msaroufim @jcaip

jcaip commented:
Thanks for the update @Diogo-V! I 'll take a more thorough pass through later tonight

Diogo-V commented:
Hi @jcaip, just saw that you did a commit getting `torch.compile` to work. Thank you!
Later tonight I will be working on getting the failing tests to pass. Is there anything else that you feel like is still missing/needs refactoring and I could take a stab at?

jcaip commented:
@Diogo-V Thanks for your help, and good job landing the op!

I took a look at your code and I do think there might be an issue in the way we hook up the marlin gemm op. I noticed for that changing the batch dimension (32, 16, 14096) for the input breaks the test, which shouldn't be the case. The compile test is also failing for me. 

I don't know exactly what's where the problem is, but the general gist is that we have `XW` implemented for marlin op but actually for linear we calculate xW', which is why I think you need to transpose `int_data` and `scales` in `from_float`. Somewhere along the lines I think we're passing in the wrong dimensions, but need to do more debugging today. 

Could you open a PR with just the marlin op (without the affine quantized tensor implementation) for now? You can git squash all your commits and then remove the AQT files with: https://stackoverflow.com/questions/12481639/remove-file-from-latest-commit. We can merge that in first, since I'll be on PTO next week. cc @jerryzh168 @msaroufim 

I think the compile debugging will be a bit more involved, so don't want you to get blocked on that. Will update if I find the bug.

Diogo-V commented:
Yeah totally! Will open the PR in a bit

I am not super acquainted with the compile process (have been learning it the past few weeks). Do you have any resources that would be good for me to take a look at so that I can help out more and get across this issues in future PRs?

This weekend, I will take another stab at this and try to get the tests to pass.

jcaip commented:
Hey @Diogo-V 

> I am not super acquainted with the compile process (have been learning it the past few weeks). Do you have any resources that would be good for me to take a look at so that I can help out more and get across this issues in future PRs?

https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html is a good tutorial. I would actually recommend not debugging with compile, until you first fix the batch size=32 eager test case, since it'll be a lot easier to be able to debug in eager. 

> This weekend, I will take another stab at this and try to get the tests to pass. I would also be happy to pass this work off to someone else if you feel like it would be too tight of a schedule for the poster.

Thanks, that would be awesome! It would definitely help to have another person debug this. Let me try to explain a bit more about what I think is going on here:

On CUDA, tensors are represented as just blobs of data, and we keep track of the strides / offsets / dimensions of the tensor separately in metadata. This is nice because we can do something like a transpose without needing to modify our existing blog of data, we just change how we access it via the strides / offsets. 

The Marlin kernel also takes in these m.k.n values, likely to make the kernel implementation easier. I think somewhere along the line we are flubbing a transpose somewhere, where we are switching up our mkn dimensions (because transpose), but still passing in the contiguous data of the non-transposed tensor. I think this mismatch is the cause of our error. 

This is especially compounded because for 2:4 spares support, the tensor cores only support the first element being sparse really, which we can deal with by using transpose properties to move the second element sparse. 
![](https://pytorch.org/assets/images/accelerating-neural-network-training/fg1.jpg)

But note how we are now returning a transposed matrix, which means it is not contiguous, hence why it is in column major [format](a8d8a07bf32b99fbe8d4a71f9d7a58e745f5c9186d836b0e163601108ceb6aafR188). I think this row/column major format mismatch, or something related is the cause of our compile / batch_size=n bug. To make sure, we can check that both the marlin output and the reference output share the same strides in our unit test: ` assert sparse_result.stride() == reference_result.stride()`

Just a tip, feel free to do whatever you feel is best: I would gradually work up from your [test_marlin_24](https://github.com/pytorch/ao/pull/621/files#diff-d183f2afc51d6a59bc70094e8f476d2468c45e415500f6eb60abad955e065156R378) test here, by writing a nn.Linear layer that outputs the same values as your reference mm implementation. 

Basically my thinking is that we go from mm unit test -> nn.Linear unit test -> e2e test. Let me know if you have any questions, I'll try to answer them the best I can. 

As far as deadlines - just FYI: 9/6 is the deadline for our torchAO announcement, with the 9/16 is the deadline for making updates to our PTC poster.

Diogo-V commented:
> https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html is a good tutorial. I would actually recommend not debugging with compile, until you first fix the batch size=32 eager test case, since it'll be a lot easier to be able to debug in eager.

Thank you very much! I will start off by tackling the batch size issue before moving onto the compile test as you suggested.

> On CUDA, tensors are represented as just blobs of data (...) unit test: `assert sparse_result.stride() == reference_result.stride()`

Yep, I agree with you on this. Thanks for sharing your insights about the current problem! I think I have a better idea of how to debug this after reading your explanation 🙂

> Just a tip, feel free to do whatever you feel is best: I would gradually work up from your [test_marlin_24](https://github.com/pytorch/ao/pull/621/files#diff-d183f2afc51d6a59bc70094e8f476d2468c45e415500f6eb60abad955e065156R378) test here, by writing a nn.Linear layer that outputs the same values as your reference mm implementation.
>
> Basically my thinking is that we go from mm unit test -> nn.Linear unit test -> e2e test. Let me know if you have any questions, I'll try to answer them the best I can.

I think that suggestion is pretty good. I will follow that flow to debug this issue.

> As far as deadlines - just FYI: 9/6 is the deadline for our torchAO announcement, with the 9/16 is the deadline for making updates to our PTC poster.

Thanks for sharing the deadlines. I was getting a bit concerned because I know that this is a time sensitive task and I don't want to risk missing an important date. I will try to surface updates and blocks as often as I can.

jerryzh168 commented on torchao/dtypes/affine_quantized_tensor.py:
```diff
@@ -56,10 +58,13 @@ def from_plain(
-        int_data, scale, zero_point = self.get_plain()
-        layout_type = self.get_layout_type()
-        return f"{self.__class__.__name__}(int_data={int_data}, scale={scale}, zero_point={zero_point}, layout_type={layout_type})"
+        # This is a hack, torch.compile tries to trace the __repr__ function which then calls `dequantize` function, causing an error.
```
can these changes be reverted now with dynamo.disable decorator?

Diogo-V replied:
yeah, will take care of that

Diogo-V commented:
Hi, this last couple of days I have been trying some things and wanted to give an update / surface some of the findings to help anyone else that might be also tackling this:

#### Update
I was able to fix a bug in how the batch_size was being handled and I seem to have been able to replicate the issue that we are facing on a smaller example. Just for debug purposes, I added a test `test_marlin_24_debug` in `test_ops.py` file. It has notes explaining which section on the larger part of the code it is trying to simulate.

#### Observation 1
I went back to the Sparse Marlin repo to find out how they handle stuff in a Linear layer.
In their tests, they create a tensor `w` with shapes `(21504, 4096)` and then transpose it into `(4096, 21504)`. In the next couple of lines, they create a Linear layer and then overwrite it's weight just before packing:
```py
w = torch.randn((21504, 4096))
w = w.t()  # Is now (4096, 21504)
...
linear = nn.Linear(4096, 21504)  # Has weight tensor of shape (21504, 4096)
linear.weight.data = w  # Is now of shape (4096, 21504), the equivalent of transposing the linear's weight
...
# q has shape (128, 43008), where 128 = 4096 / 2 / 16 and 43008 = 21504 * 16 / 8
q = pack(linear, scales)  
```
After packing, the shapes are a transformation of the original transposed tensor. This seems to indicate that the transpose inside the `from_plain` method might be necessary.

#### Observation 2
In the Sparse Marlin repo, I also tried to get the output metadata and, even though, in their documentation, it is mentioned that the output is in column-major format, the strides are `(21504, 1)` which seems to indicate that it is in row-major format and when calling `is_contiguous()` on the output tensor, the result is also `True`. Is there an instance that I am not aware of where it is possible for a tensor to be in column-major format and have the above mentioned strides?

#### Observation 3
Still in the Sparse Marlin repo, I adapted their `replace_linear` function to replace the linear layers with the implementation of `Layer_2_4` from their repo as well as perform all the changes and computations that they have in their test examples to get it to work. I used a TinyLlama model and the results were not great:
<img width="1179" alt="image" src="https://github.com/user-attachments/assets/e4c6f895-3f1e-49bd-980d-04b050df9695">
I made sure to check that the weights from my adapted code matched and generated the same weights from the tests.

#### Considerations
I am still not sure where we might be failing but I am going to continue on this. Does anyone have any more insights or ideas as to what might be causing this issue?

jerryzh168 commented:
@Diogo-V thanks for the update, I feel the suggested order of `mm unit test -> nn.Linear unit test -> e2e test.` makes sense, I'm wondering if you can run https://github.com/neuralmagic/nm-vllm/blob/9daca33a6fdc429802f448e1ea71630c996c9740/vllm/model_executor/layers/quantization/marlin.py#L232 and print the intermediate results and compare with the implementation in torchao to narrow down the problem?

Diogo-V commented:
> @Diogo-V thanks for the update, I feel the suggested order of `mm unit test -> nn.Linear unit test -> e2e test.` makes sense, I'm wondering if you can run https://github.com/neuralmagic/nm-vllm/blob/9daca33a6fdc429802f448e1ea71630c996c9740/vllm/model_executor/layers/quantization/marlin.py#L232 and print the intermediate results and compare with the implementation in torchao to narrow down the problem?

Yeah absolutely! That seems a good thing to try. Will let you know of the results once I have them

Diogo-V commented:
Hi @jerryzh168, so I believe I have been able to fix the problem that we were having in the code and now the model outputs reasonable responses and seems to be correct:
<img width="457" alt="image" src="https://github.com/user-attachments/assets/91bcaf79-ff63-46ea-9bb7-aa58f8f38822">

After the previous debug that I did, I was able to narrow down the issue to the function `inject_24`. The bug was happening because applying that function to a transposed matrix and then transposing the output, does not yield the same results as applying the function to a non-transposed matrix:

```py
x = torch.randn(128, 128)
y = copy.deepcopy(x)

x_24 = inject_24(x, *x.shape)

y_t = y.t()
y_t_24 = inject_24(y_t, *y_t.shape)
y_24 = y_t_24.t()

assert torch.equal(x_24, y_24)  # This will raise an expection
```

With this, we might just be missing getting the `torch.compile` to work without using the `@torch._dynamo.disable` decorators. I will take a stab at it during the weekend.

Let me know if think there is anything that might need to be refactored. I would be more than happy to do so.

jerryzh168 commented on torchao/quantization/utils.py:
```diff
@@ -362,8 +362,7 @@ def groupwise_affine_quantize_tensor_from_qparams(
             int_data = int_data.cpu()
-        int_data = (int_data[::, ::2] << 4 | int_data[::, 1::2]).to(torch.uint8)
-        if int_data_device_type == 'mps':
+            int_data = (int_data[::, ::2] << 4 | int_data[::, 1::2]).to(torch.uint8)
```
why is this change needed? I thought the previous version is going to handle devices other than "mps" as well

Diogo-V replied:
I am not entirely sure about it as that changed was done by @jcaip. He might have a bit more context regarding this change

jcaip replied:
I think this is just a typo on my end.

jerryzh168 commented:
That's great @Diogo-V, glad to see the progress here, only thing is I'm thinking of refactoring some of the dispatch code in affine quantized tensor to allow people to define their layout and quantized linear dispatch in separate files for better code structure, but it's not ready yet, will let you know when it's ready.

jcaip commented:
@Diogo-V Nice catch! That definitely seems like the source of the issue. I've been on PTO the last week, but getting back into this now. 

cc @jerryzh168 do you know when you expect that to be ready by? Just asking because I have some BSR + int8 work that I want to land and not sure if I should hold off on that as well. Thanks for your help guiding diogo too while I've been out.

jerryzh168 commented:
@jcaip yeah no problem, the refactor PR is here: https://github.com/pytorch/ao/pull/783 feel free to stamp

Diogo-V commented:
@jcaip - Awesome, glad you are back and hope you had a great PTO :)

To give a bit on an update on this: I had a busier weekend than I was expecting but still managed to take a look at what is currently happening. When running the example that compiles the forward pass from the file `wip_test_llama2.py`, an exception was raised on this [line](https://github.com/Diogo-V/ao/blob/68b3b71205a60af799eb780b2f87af06951845b4/torchao/dtypes/affine_quantized_tensor.py#L1293) with the message saying that the `weight_tensor` was a `FakeTensor` and did not have the property `layout_type`. 

That was as far as I got but I am planning on taking another look at it tomorrow evening.

Let me know if you have any insights into what might be happening here.

jerryzh168 commented:
@Diogo-V the error might be related to version of pytorch, can you try with pytorch nightly version? you can first `pip uninstall torch` and then reinstall with https://pytorch.org/get-started/locally/

Diogo-V commented:
@jerryzh168 - After some debugging, I was able to fix most of the issues that were happening when `torch.compile` was used. There seems to be just one last problem that, once fixed, will let us run using `torch.compile`. 

The issue is on [this](https://github.com/Diogo-V/ao/blob/d4fb47fa20fcbcbba6ea998eb24f2b748327c79a/torchao/sparsity/marlin/__init__.py#L269) line and the raised exception is `RuntimeError: .numpy() is not supported for tensor subclasses.`. This seems to be happening because a `FakeTensor` is being passed here and it does not support numpy.

However, a similar action is also done in [this](https://github.com/Diogo-V/ao/blob/d4fb47fa20fcbcbba6ea998eb24f2b748327c79a/torchao/sparsity/marlin/__init__.py#L237) line but no errors are thrown which left me a bit confused.

So, I have the below questions: 
  1. Is there a function and/or approach that I need to use to get dynamo to work with this section of numpy code?
  2. Do you have any insights/tips on how to approach this?

jerryzh168 commented:
> @jerryzh168 - After some debugging, I was able to fix most of the issues that were happening when `torch.compile` was used. There seems to be just one last problem that, once fixed, will let us run using `torch.compile`.
> 
> The issue is on [this](https://github.com/Diogo-V/ao/blob/d4fb47fa20fcbcbba6ea998eb24f2b748327c79a/torchao/sparsity/marlin/__init__.py#L269) line and the raised exception is `RuntimeError: .numpy() is not supported for tensor subclasses.`. This seems to be happening because a `FakeTensor` is being passed here and it does not support numpy.
> 
> However, a similar action is also done in [this](https://github.com/Diogo-V/ao/blob/d4fb47fa20fcbcbba6ea998eb24f2b748327c79a/torchao/sparsity/marlin/__init__.py#L237) line but no errors are thrown which left me a bit confused.
> 
> So, I have the below questions:
> 
> 1. Is there a function and/or approach that I need to use to get dynamo to work with this section of numpy code?
> 2. Do you have any insights/tips on how to approach this?

this is an error on tensor subclasses, I'm wondering why `q_packed.cpu()` is a tensor subclass? if it's because of FakeTensor I feel we just have to workaround this by not using numpy there. I'd suggest to rewrite `_from_marlin_weights` with native pytorch APIs instead of using numpy.

> Is there a function and/or approach that I need to use to get dynamo to work with this section of numpy code?

it seems numpy with pytorch native tensor works: https://pytorch.org/blog/compiling-numpy-code/ the problem is just tensor subclass + numpy

Diogo-V commented:
@jerryzh168 - I was able to solve the issue and so, I think the PR is done now. Will mark it for review.

jerryzh168 commented on test/sparsity/test_marlin.py:
```diff
@@ -0,0 +1,132 @@
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="Need CUDA available")
+    def test_quant_sparse_marlin_layout_eager(self):
+        torch.manual_seed(0)
```
nit: maybe this can be moved to setup like in https://github.com/pytorch/pytorch/blob/dbeb8a1691a9dba4bf2a10566fd55705d67c24a9/torch/testing/_internal/common_quantization.py#L533

Diogo-V replied:
Refactored the test to use the `setUp` method

jerryzh168 commented on torchao/sparsity/marlin/utils.py:
```diff
@@ -93,22 +92,17 @@ def reverse_marlin_permute_weights(
-    
-    As a result of this reordering, the vector loads inside the kernel will get the data as it is needed for tensor-core
-    (without the need to use ldmatrix instructions)
+    with the tensor-core format.
```
why the docs are removed?

Diogo-V replied:
This was a miss click. Thanks for catching it

jerryzh168 commented on test/sparsity/test_marlin.py:
```diff
@@ -0,0 +1,132 @@
+        # Inject 2:4 sparsity mask
+        w_24, _ = inject_24(w, *w.shape)
+
+        # Quantize weights 
+        w_24 = w_24.reshape((-1, group_size, size_n))
+        w_24 = w_24.permute(1, 0, 2)
+        w_24 = w_24.reshape((group_size, -1))
+
+        # Compute scale for each group
+        scales = torch.max(torch.abs(w_24), 0, keepdim=True)[0]
+        scales *= 2 / max_q_val  # 2 => symmetric
```
these seems to be complicated, should we add a test to make sure the quantized weight and calculate scale matches the result of using affine quantized tensor quant primitives in `int4_weight_only` with the `MarlinSparseLayoutType`?

Diogo-V replied:
Yeah you are right. Just updated the test to instead use the quantization APIs from `int4_weight_only`

jerryzh168 replied:
if this becomes a duplicate with test_sparse_api, maybe we can remove this one I feel

jerryzh168 reviewed (commented):
looks good overall, thanks for working on this!

left some inline comments and I think we can add a test for this in https://github.com/pytorch/ao/blob/e05635eb25903e8d623c1c88aad44339c41789c3/test/sparsity/test_sparse_api.py#L29-L52, and also add the API to README, maybe in https://github.com/pytorch/ao/tree/main/torchao/sparsity, cc @jcaip looks like current README does not have a section talking about APIs yet

can be next PR:
we can also do e2e performance and accuracy eval on sam model with https://github.com/pytorch/ao/tree/main/torchao/_models/sam and report the numbers as well I think, cc @jcaip will it be useful to have a table like https://github.com/pytorch/ao/tree/main/torchao/quantization#benchmarks for sparsity?

Diogo-V commented:
I had to add the below decorator to the test `test_quant_sparse_marlin_layout_compile`
```py
@pytest.mark.skipif(not TORCH_VERSION_AT_LEAST_2_5, reason="Needs PyTorch 2.5")
```
because it was raising this exception (you can observe it on the CI):
 
![image](https://github.com/user-attachments/assets/161b9858-77e7-4987-9747-450c2e8dbcb8)

I had previous gotten around this issue by installing the nightly version as you had suggested. Is this the correct thing to do or should I try something else?

Diogo-V commented:
@jerryzh168 - Just addressed your review and I have a question:

You suggested adding a test in the file `test_sparse_api.py` that would be similar to those there. Wouldn't that be repeating the tests implemented in `test_marlin.py`?

Was this the kind of test that you were expecting to add on that file or did you have something else on your mind?
```py
    def test_sparse_marlin(self):
        input = torch.rand((256, 256)).half().cuda()
        model = (
            nn.Sequential(
                nn.Linear(256, 1024),
                nn.Linear(1024, 256),
            )
            .half()
            .cuda()
        )

        apply_fake_sparsity(model)
        model_copy = copy.deepcopy(model)

        # Quantized
        quantize_(model_copy.bfloat16(), int4_weight_only())
        dense_result = model_copy(input.bfloat16()).half()

        # Sparse + quantized
        quantize_(model, int4_weight_only(layout_type=MarlinSparseLayoutType()))
        sparse_result = model(input)

        assert torch.allclose(dense_result, sparse_result, atol=3e-1), "Results are not close"
```

Let me know if there is anything else that needs to be updated!

jcaip commented:
I tried runnning this with SAM but was running into a CUDA graph error, however with the 24_sparse_hf_example, I'm seeing:
```
Tokens/second: 134.497 for 2:4 sparse marlin 
Tokens/second: 111.182 for our int4 tinygemm version 
Tokens/second: ~80 for 2:4 sparse
Tokens/second: ~75 for compile baseline
```
🚀 🚀 🚀  cc @Diogo-V 

I'll write this up and update the README @jerryzh168 for BE day tomorrow. It would be best if we could hook up into @HDCharles LLaMa benchmarks for consistency.

jerryzh168 commented:
> I had previous gotten around this issue by installing the nightly version as you had suggested. Is this the correct thing to do or should I try something else?

only test this in nightly sounds good if that's the issue

jerryzh168 commented:
> @jerryzh168 - Just addressed your review and I have a question:
> 
> You suggested adding a test in the file `test_sparse_api.py` that would be similar to those there. Wouldn't that be repeating the tests implemented in `test_marlin.py`?
> 
> Was this the kind of test that you were expecting to add on that file or did you have something else on your mind?
> 
> ```python
>     def test_sparse_marlin(self):
>         input = torch.rand((256, 256)).half().cuda()
>         model = (
>             nn.Sequential(
>                 nn.Linear(256, 1024),
>                 nn.Linear(1024, 256),
>             )
>             .half()
>             .cuda()
>         )
> 
>         apply_fake_sparsity(model)
>         model_copy = copy.deepcopy(model)
> 
>         # Quantized
>         quantize_(model_copy.bfloat16(), int4_weight_only())
>         dense_result = model_copy(input.bfloat16()).half()
> 
>         # Sparse + quantized
>         quantize_(model, int4_weight_only(layout_type=MarlinSparseLayoutType()))
>         sparse_result = model(input)
> 
>         assert torch.allclose(dense_result, sparse_result, atol=3e-1), "Results are not close"
> ```
> 
> Let me know if there is anything else that needs to be updated!

yeah looks good to me

jerryzh168 reviewed (approved):
looks good to me, please move the test we are discussing to test_sparse_api.py and make sure CI passes before merging

jerryzh168 commented on torchao/dtypes/affine_quantized_tensor.py:
```diff
@@ -1219,6 +1413,47 @@ def _linear_fp_act_fp8_weight_impl(
     ).reshape(out_shape)
 
 
+def _linear_fp_act_int4_weight_sparse_marlin_check(input_tensor, weight_tensor, bias):
```
cc @jcaip now we support registering layout type and these things in a separate file, maybe we can create a folder for all sparse layouts and move these

jcaip replied:
Yeah, that sounds like a good idea.

Diogo-V commented:
@jerryzh168 - Should be good to go now :)

jcaip commented:
Perfect, let me add my benchmarking updates to this branch so it's easier for Andrew to cherry-pick the whole thing. 

BTW I'm seeing 226 tok/s for LLama3 - a 25% increase from 180 tok/s for our tiny gemm int4 kernel. 
The perplexity is bad, we'll have to use the neuralmagic checkpoint for a fair comparison, but the weights are in a different format than the LLama3 checkpoint and I don't want to block this on that.

Diogo-V commented:
@jcaip - Yeah, I also noticed that sometimes the model's outputs were not super great. My simple sanity check was comparing against a `quantize_(fake_sparsity(model), int4_wo(layout=Marlin...))` version and they were equal most of the time.

Let me know if there is anything else that I might be of help to get this across today and feel free to ping me on discord (Diogo-V on #cuda-mode) if you run into any issues!

jcaip commented:
Yes, for context, I am seeing the same text generation output on the hf example with the neuralmagic 2:4 sparse model, when I run `quantize_(model, int4_wo())` and `quantize_(model, int4_wo(layout=SparseMarlinLayout()))` so I don't think it's an implementation issue. I'm guessing we'll may need to also use GPTQ as well for the best accuracy numbers. 

But I'm very pleased with the speedup you've achieved (+25%), thats a great result :)

Diogo-V commented:
I wouldn't mind taking a look at adding support for GPTQ with Sparse Marlin if you feel like it would be a good thing to work on afterwards. I would have some time for it after the 21st of September.
